File: terminator-application/src/main/java/io/quartic/terminator/CatalogueWatcher.java
package io.quartic.terminator;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Sets;
import io.quartic.catalogue.api.DatasetConfig;
import io.quartic.catalogue.api.DatasetId;
import io.quartic.catalogue.api.TerminationId;
import io.quartic.catalogue.api.TerminatorDatasetLocator;
import io.quartic.common.client.WebsocketListener;
import org.immutables.value.Value;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import rx.Subscription;

import java.util.Map;
import java.util.Set;

import static io.quartic.weyl.common.serdes.ObjectMappers.OBJECT_MAPPER;
import static java.util.stream.Collectors.toCollection;

@Value.Immutable
public abstract class CatalogueWatcher implements AutoCloseable {
    private static final Logger LOG = LoggerFactory.getLogger(CatalogueWatcher.class);

    public static CatalogueWatcher of(WebsocketListener.Factory listenerFactory) {
        return ImmutableCatalogueWatcher.of(listenerFactory);
    }

    private Subscription subscription = null;
    private final Set<TerminationId> terminationIds = Sets.newHashSet();

    @Value.Parameter
    protected abstract WebsocketListener.Factory listenerFactory();

    @Value.Lazy
    protected WebsocketListener<Map<DatasetId, DatasetConfig>> listener() {
        return listenerFactory().create(OBJECT_MAPPER.getTypeFactory().constructMapType(Map.class, DatasetId.class, DatasetConfig.class));
    }

    public void start() {
        subscription = listener()
                .observable()
                .subscribe(this::update);
    }

    @Override
    public void close() throws Exception {
        if (subscription != null) {
            subscription.unsubscribe();
            subscription = null;
        }
    }

    private void update(Map<DatasetId, DatasetConfig> datasets) {
        LOG.info("Received catalogue update");
        synchronized (terminationIds) {
            terminationIds.clear();
            datasets.values().stream()
                    .filter(config -> config.locator() instanceof TerminatorDatasetLocator)
                    .map(config -> ((TerminatorDatasetLocator) config.locator()).id())
                    .collect(toCollection(() -> terminationIds));
        }
    }

    public Set<TerminationId> terminationIds() {
        synchronized (terminationIds) {
            return ImmutableSet.copyOf(terminationIds);
        }
    }
}

File: terminator-application/src/main/kotlin/io/quartic/terminator/CatalogueWatcher.kt
package io.quartic.terminator

import io.quartic.catalogue.api.DatasetConfig
import io.quartic.catalogue.api.DatasetId
import io.quartic.catalogue.api.TerminationId
import io.quartic.catalogue.api.TerminatorDatasetLocator
import io.quartic.common.websocket.WebsocketListener
import io.quartic.common.logging.logger
import io.quartic.common.serdes.objectMapper
import rx.Subscription

class CatalogueWatcher(private val listenerFactory: WebsocketListener.Factory) : AutoCloseable {
    private val LOG by logger()

    private var subscription: Subscription? = null

    private val _terminationIds = mutableSetOf<TerminationId>()
    val terminationIds: Set<TerminationId> get() {
        synchronized(_terminationIds) {
            return _terminationIds.toSet()
        }
    }

    private val listener: WebsocketListener<Map<DatasetId, DatasetConfig>> by lazy {
        listenerFactory.create<Map<DatasetId, DatasetConfig>>(
                objectMapper().typeFactory.constructMapType(Map::class.java, DatasetId::class.java, DatasetConfig::class.java)
        )
    }

    fun start() {
        subscription = listener.observable.subscribe({ this.update(it) })
    }

    override fun close() {
        if (subscription != null) {
            subscription!!.unsubscribe()
            subscription = null
        }
    }

    private fun update(datasets: Map<DatasetId, DatasetConfig>) {
        LOG.info("Received catalogue update")
        synchronized(_terminationIds) {
            _terminationIds.clear()
            _terminationIds.addAll(
                    datasets.values
                            .filter({ config -> config.locator() is TerminatorDatasetLocator })
                            .map({ config -> (config.locator() as TerminatorDatasetLocator).id() })
            )
        }
    }
}
